### PR TITLE
CDAP-7113 Close stream files after some duration

### PR DESCRIPTION
**Pull request by anwar6953**
Previously, stream files are only closed if its old enough and when the next event to it came.
This can mean that if no next event comes in, it won't be closed.

Now, have a timer to close the file after some duration.

https://issues.cask.co/browse/CDAP-7113
http://builds.cask.co/browse/CDAP-RUT135-1

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java [hunk 1]
```diff
@@ -64,8 +69,6 @@
   private final long partitionDuration;
   private TimePartition timePartition = new TimePartition(-1L);
 
-  // TODO: Add a timer task to close file after duration has passed even there is no writer.
```
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java:67]: I'm not sure what it means by "even there is no writer".
There will always be a writer, I think. There just may be no new events.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java [hunk 2]
```diff
@@ -162,7 +167,25 @@ public int hashCode() {
...
+            // NOTE: this doesn't remove it from PartitionedFileWriter map of writers
+            // it will get removed from there when the next event comes
+            streamDataFileWriter.close();
+          } catch (IOException e) {
+            Throwables.propagate(e);
+          }
+        }
+        // add a 5 minute buffer, to ensure that all events for this time partition are enqueued to the writer
+        // before closing it
+      }, partitionDuration + TimeUnit.MINUTES.toMillis(5));
```
Comment (gokulavasan) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: What is the guarantee that all events will be enqueued within 5 mins?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java [hunk 3]
```diff
@@ -162,7 +167,25 @@ public int hashCode() {
...
+
+      timer.schedule(new TimerTask() {
+        @Override
+        public void run() {
+          try {
+            // NOTE: this doesn't remove it from PartitionedFileWriter map of writers
+            // it will get removed from there when the next event comes
+            streamDataFileWriter.close();
+          } catch (IOException e) {
+            Throwables.propagate(e);
```
Comment (gokulavasan) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: Should we propagate this exception if we are not able to close the stream files?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java (hunk 2 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: Enqueuing events to the writer (after computing a time for that event) is a very quick process.
Much of the delay in writing will be after its already enqueued to that writer.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java [hunk 4]
```diff
@@ -152,17 +157,37 @@ public int hashCode() {
...
         throw new IOException("Failed to create new file at " + eventFile + " and " + indexFile);
       }
 
       LOG.debug("New stream file created at {}", eventFile);
-      return new StreamDataFileWriter(createOutputSupplier(eventFile), createOutputSupplier(indexFile), indexInterval);
+
+      final StreamDataFileWriter streamDataFileWriter =
+        new StreamDataFileWriter(createOutputSupplier(eventFile), createOutputSupplier(indexFile), indexInterval);
+
+      timer.schedule(new TimerTask() {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: It's better not to do it on the stream data file directly.

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: Instead, the `closePartitionWriter` should be used. You can simply check against the current `timePartition`.

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: And also, there can be a lot of threads running around (one per stream id).

Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: There's already logic using `closePartitionerWriter`, checking against the current `timePartition`.
However, with that, the files will remain open, if there's no new events.

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: I have some idea how to do it. Basically at the ConcurrentStreamWriter level, let's say we have one thread that runs once every n minute, which is the the smallest partition time of all streams in the eventQueues map (you can on every run, determine what's the smallest and schedule it's next run time).

Then we add a `refresh()` method to the `EventQueue`. What it does is to become leader of that queue (see `tryWrite` method) and call `refresh` method on the current `FileWriter` (maybe define an interface for that method so that it won't mess with the FileWriter interface). The implementation of the `TimePartitionedStreamFileWriter` will implement that interface. Upon refresh is being called, it will close the file if the current partition time is passed.

Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: That seems a bit difficult to do because at the `ConcurrentStreamWriter` level, the FileWriter is not guaranteed to be a `PartitionedFileWriter` nor a `TimePartitionedStreamFileWriter`, so we can't really check which is the smallest partition time.

Also, since its operating on the `FileWriter` interface, seems like we need to add this `refresh` method to the `FileWriter` interface.

Alternatively, maybe we can share a single thread pool across all the streams by injecting the executor as singleton via Guice to all the different `TimePartitionedStreamFileWriter`s.

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: The partition time is from the stream config, which is accessible from the concurrent stream writer

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/file/FileWriter.java [hunk 5]
```diff
@@ -42,4 +42,11 @@
    * @throws IOException if fail to append
    */
   void appendAll(Iterator<? extends T> events) throws IOException;
+
+  /**
+   * Refreshes the FileWriter.
+   *
+   * @throws IOException if fail to refresh
+   */
+  void refresh() throws IOException;
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/file/FileWriter.java]: Don't add it to this interface. Create a separate interface

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java (hunk 4 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: That's the partitionDuration, but not the partitionTime of the writer in the queue, right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java [hunk 6]
```diff
@@ -112,6 +115,29 @@
     this.generationWatched = Sets.newHashSet();
     this.cancellables = Lists.newArrayList();
     this.createLock = new ReentrantLock();
+    this.eventQueueRefreshService = scheduleWriterRefresh();
+  }
+
+  public AbstractScheduledService scheduleWriterRefresh() {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java]: Return the Service interface

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/file/FileWriter.java (hunk 5 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/file/FileWriter.java]: `FileWriter` will still need to implement that other interface, right?
Because `ConcurrentStreamWriter` deals with `FileWriter` interface, not the concrete class.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java [hunk 7]
```diff
@@ -536,10 +571,28 @@ boolean tryWrite() {
...
+     *
+     * @return true if become the writer leader and performed the refresh, false otherwise.
+     */
+    boolean refresh() throws IOException {
+      if (!writerFlag.compareAndSet(false, true)) {
+        return false;
+      }
+
+      try {
+        this.getFileWriter().refresh();
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java]: Use instance of here so that you don't have to make every file writer carries a refresh method

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java [hunk 8]
```diff
@@ -536,10 +571,28 @@ boolean tryWrite() {
     }
 
     /**
+     * Attempts to refresh the underlying FileWriter.
+     *
+     * @return true if become the writer leader and performed the refresh, false otherwise.
+     */
+    boolean refresh() throws IOException {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java]: You are not using the return value, why not return void?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java [hunk 9]
```diff
@@ -321,14 +349,21 @@ public void deleted(Id.Stream streamId) {
      * @return A {@link FileWriter} for writing {@link StreamEvent} to the given stream
      * @throws IOException if failed to create the file writer
      */
-    FileWriter<StreamEvent> create(Id.Stream streamId) throws Exception {
+    private FileWriter<StreamEvent> create(Id.Stream streamId) throws IOException {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: Why need to change this method?

Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: it didn't need to throw `Exception`, only `IOException`. simplified the exception handling of the calling code.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java (hunk 4 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: ? The partition start time are calculated from the stream config.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/file/FileWriter.java (hunk 5 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/file/FileWriter.java]: No. You can use instance of in the concurrent stream writer

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java (hunk 9 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: Did you simplied the caller code? I don't see it in the diff

Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: I made `getFileWriter` return only `IOException` now.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java [hunk 10]
```diff
@@ -536,10 +573,31 @@ boolean tryWrite() {
...
+     * @return true if become the writer leader and performed the refresh, false otherwise.
+     */
+    boolean refresh() throws IOException {
+      if (!writerFlag.compareAndSet(false, true)) {
+        return false;
+      }
+
+      try {
+        FileWriter fileWriter = this.getFileWriter();
+        if (fileWriter instanceof TimePartitionedStreamFileWriter) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java]: That's not a good one. It's too specific. introduce an interface, like Refreshable.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java (hunk 9 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: But I don't see code change on the caller of getFileWriter. The refresh method can throw Exception. I don't see why it has to limit to IOException. The concept of Refreshable is pretty generic

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: Also, we should design a method signature based on how it is called. It should based on what it represents

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/file/FileWriter.java [hunk 11]
```diff
@@ -42,4 +42,5 @@
    * @throws IOException if fail to append
    */
   void appendAll(Iterator<? extends T> events) throws IOException;
+
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/file/FileWriter.java]: Unnecessary changr

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/file/PartitionedFileWriter.java [hunk 12]
```diff
@@ -13,6 +13,7 @@
  * License for the specific language governing permissions and limitations under
  * the License.
  */
+
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/file/PartitionedFileWriter.java]: Unnecessary change

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/file/FileWriter.java (hunk 11 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/file/FileWriter.java]: its leftover from when I removed the method from this interface. I'll remove it.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java [hunk 13]
```diff
@@ -152,10 +173,10 @@ public int hashCode() {
 
       // Create the event and index file with the next sequence id
       int fileSequence = StreamUtils.getNextSequenceId(partitionDirectory, fileNamePrefix);
-      Location eventFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
-                                                            fileSequence, StreamFileType.EVENT);
-      Location indexFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
-                                                            fileSequence, StreamFileType.INDEX);
+      final Location eventFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: Unneeded change

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java [hunk 14]
```diff
@@ -88,6 +89,26 @@ protected void partitionChanged(TimePartition oldPartition, TimePartition newPar
...
+   *
+   * @throws IOException if fail to refresh
+   */
+  public void refresh() throws Exception {
+    if (timePartition.getStartTimestamp() == -1) {
+      // timePartition hasn'scheduleWriterRefresh been overridden yet
+      return;
+    }
+
+    // Add a 5 minute buffer, to ensure that all events for this time partition are enqueued to the writer
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java]: Actually we don't need to add time. This is because this call is protected by the leader lock (notice this class is marked with not thread safe) and there won't be any unfinished write

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java [hunk 15]
```diff
@@ -585,6 +641,60 @@ private void doClose() {
...
+      }
+
+      @Override
+      public void flush() throws IOException {
+        writer.flush();
+      }
+
+      @Override
+      public void refresh() throws Exception {
+        // TODO: cache the value
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java]: Cache what value? Checking instanceof is a very fast instruction

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java (hunk 9 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: Undo this change

Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: The only checked exception that this method will throw is `IOException`, though

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: I mean undo the change in this method. Impersonator can throw other exception

Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: Impersonator `doAs` method simply throws `Exception`, and the only checked exception it throws (besides `IOException`) is whatever the `Callable` throws. However, `Callable` is blanket `throws Exception`

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: I see. On a separate note, we can improve the doAs to throw the same exception in the signature as the task to perform

Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: I see. I didn't think about Generics for `throws`, but I see its possible now.

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: To do so, instead of using Callable, introduce a new interface, say `Task<V, E extends Exception>`, with a method `V perform() throws E`. Then the doAs signature becomes `<V, E extends Exception> doAs(NamespaceId I'd, Task<V, E> task)`

Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java:354]: Filed a JIRA for 4.0:
https://issues.cask.co/browse/CDAP-7130

**Comment by chtyim** [comment]
LGTM. Make sure you tested it on the cluster. You might want to add a debug log to the refresh method of the file writer and log when the file is closed.

**Comment by anwar6953** [comment]
Added debug log.
Will test on cluster, that behavior is as expected.

**Comment by anwar6953** [comment]
Thanks for the review @chtyim .

**Comment by anwar6953** [comment]
Squashed commits. Couldn't trigger a build, because bamboo was down:

image

**Comment by sreevatsanraman** [comment]
Builds are running now. http://builds.cask.co/browse/CDAP-RUT135 Bamboo was doing a backup at that time so there could have been unavailability at that time.

**Comment by sreevatsanraman** [comment]
Build passed. merging.
